// jcatalog-i18n/lib/utils/I18nManager.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _converters = require('../converters');

var _constants = require('./constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var deepMerge = function deepMerge(object, source) {
  return _lodash2.default.mergeWith(object, source, function (objValue, srcValue) {
    if (_lodash2.default.isObject(objValue) && srcValue) {
      return deepMerge(objValue, srcValue);
    }
    return undefined;
  });
};

/**
 * Performs initialization and handling of intl data.
 * It supports intl data in React Intl format.
 *
 * @author Alexander Frolov
 */

var I18nManager = function () {
  /**
   * Creates and initialize new manager instance.
   * Where:
   *- locale is current locale
   *- intDatas is an list with messages for different locales.
   *- formatInfos is format pattern data
   *- defaultLocale - fallback locale, 'en' by default
   * (see React Intl Data format)
   */
  function I18nManager(locale, intlDatas) {
    var formatInfos = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
    var defaultLocale = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 'en';

    _classCallCheck(this, I18nManager);

    _initialiseProps.call(this);

    this._intlData = { locales: [locale], messages: {} };

    this._intlDatas = [this._intlData];
    this._components = [];

    // current locale
    this.locale = locale;
    this.defaultLocale = defaultLocale;
    if (intlDatas) {
      this.register('default', intlDatas);
    }

    this._formatInfos = formatInfos;
    if (formatInfos && formatInfos[locale]) {
      this._formatInfo = formatInfos[locale];
    } else {
      this._formatInfo = _constants.DEFAULT_FORMAT_INFO;
    }

    var numberGroupingSeparator = null;
    if (this._formatInfo.numberGroupingSeparatorUse) {
      numberGroupingSeparator = this._formatInfo.numberGroupingSeparator;
    }

    this._dateConverter = new _converters.DateConverter(this._formatInfo.datePattern);
    this._dateTimeConverter = new _converters.DateConverter(this._formatInfo.dateTimePattern);
    this._decimalNumberConverter = new _converters.NumberConverter(this._formatInfo.numberPattern, numberGroupingSeparator, this._formatInfo.numberDecimalSeparator, this._formatInfo.numberDecimalSeparatorUseAlways);
    this._numberConverter = new _converters.NumberConverter(this._formatInfo.integerPattern, numberGroupingSeparator, null, this._formatInfo.numberDecimalSeparatorUseAlways);
  }

  _createClass(I18nManager, [{
    key: '_getMessageBundleForLocale',


    /**
    * Searches for a budle that locales collection containes argument-locale
    */
    value: function _getMessageBundleForLocale(locale) {
      return _lodash2.default.find(this._intlDatas, function (storedIntlData) {
        return _lodash2.default.indexOf(storedIntlData.locales, locale) !== -1;
      });
    }

    /**
    * Returns fallback locale with the next logic: fr-FR->fr->en
    */

  }, {
    key: '_getFallbackLocale',
    value: function _getFallbackLocale() {
      if (this.locale.indexOf('-') !== -1) {
        return this.locale.substring(0, this.locale.indexOf('-'));
      }

      return this.defaultLocale;
    }
  }, {
    key: 'dateFormat',
    get: function get() {
      return this._formatInfo.datePattern;
    }
  }]);

  return I18nManager;
}();

var _initialiseProps = function _initialiseProps() {
  var _this = this;

  this.register = function (component, intlDatas) {
    if (_this._components.indexOf(component) < 0) {
      (function () {
        // function to check if locales have common elements
        var intersects = function intersects(l1, l2) {
          for (var i = 0; i < l1.length; i++) {
            if (l2.indexOf(l1[i]) >= 0) {
              return true;
            }
          }
          return false;
        };

        var that = _this;

        // processing collection of bundle arguments:
        // in the next commnts 'bundle' means js object with the next notation:
        // {locales: [], messages: {}}
        intlDatas.forEach(function (intlData) {
          // searching for already existing bundle with 'similliar' locale collection
          var indexToExtend = _lodash2.default.findIndex(that._intlDatas, function (storedIntlData) {
            return intersects(storedIntlData.locales, intlData.locales);
          });

          // if we find bundle with locales that intersect with the exteernal one
          // we merge merge map of their messages and unite locales-collections
          if (indexToExtend !== -1) {
            that._intlDatas[indexToExtend] = _lodash2.default.extend({}, { locales: _lodash2.default.union(that._intlDatas[indexToExtend].locales, intlData.locales) }, {
              messages: deepMerge(that._intlDatas[indexToExtend].messages, intlData.messages)
            });
          } else {
            // otherwise we save this bundle to the internal collection of bundles
            that._intlDatas.push(intlData);
          }
        });
        _this._intlData = _this._getMessageBundleForLocale(_this.locale);
        _this._components.push(component);
      })();
    }

    return _this;
  };

  this.getMessage = function (path) {
    var args = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    var messages = _this._intlData.messages;
    var pathParts = path.split('.');

    var message = undefined;
    try {
      message = pathParts.reduce(function (obj, pathPart) {
        return obj[pathPart];
      }, messages);
    } catch (e) {
      // ignore and go next
    }
    if (message === undefined) {
      try {
        message = pathParts.reduce(function (obj, pathPart) {
          return obj[pathPart];
        }, _this._getMessageBundleForLocale(_this._getFallbackLocale()).messages);
      } catch (e) {
        // ignore and go next
      }
    }
    if (message === undefined) {
      try {
        message = pathParts.reduce(function (obj, pathPart) {
          return obj[pathPart];
        }, _this._getMessageBundleForLocale(_this.defaultLocale).messages);
      } catch (eee) {
        // ignore and go next
      }
    }
    if (message === undefined) {
      message = path;
    }

    // this check covers use case of object message,
    // f.e. message === { test: 'test component', format: 'min={min}, max={max}' }
    if (!_lodash2.default.isString(message)) {
      return path;
    }

    _lodash2.default.each(_lodash2.default.keys(args), function (param) {
      var paramValue = args[param];

      if (paramValue !== null && paramValue !== undefined) {
        message = message.replace('{' + param + '}', paramValue.toString());
      }
    });
    return message;
  };

  this.formatDate = function (date) {
    return _this._dateConverter.valueToString(date);
  };

  this.formatDateTime = function (date) {
    return _this._dateTimeConverter.valueToString(date);
  };

  this.formatDecimalNumber = function (number) {
    return _this._decimalNumberConverter.valueToString(number);
  };

  this.formatNumber = function (number) {
    return _this._numberConverter.valueToString(number);
  };

  this.parseDate = function (string) {
    return _this._dateConverter.stringToValue(string);
  };

  this.parseDecimalNumber = function (string) {
    return _this._decimalNumberConverter.stringToValue(string);
  };

  this.parseNumber = function (string) {
    return _this._numberConverter.stringToValue(string);
  };
};

exports.default = I18nManager;
